spec/lib/patches/aws-sdk-core/seahorse_client_net_http_pool_patch_spec.rb

```diff
@@ -0,0 +1,45 @@
+require 'aws-sdk'
+require 'patches/aws-sdk-core/seahorse_client_net_http_pool_patch'
+
+describe Seahorse::Client::NetHttp::ConnectionPool do
+  describe "#start_session (monkey patched)" do
+    let(:patch_file)      { File.join(%w[lib patches aws-sdk-core seahorse_client_net_http_pool_patch.rb]) }
+    let(:patch_filepath)  { ManageIQ::Providers::Amazon::Engine.root.join(patch_file).to_s }
+    let(:source_location) { subject.method(:start_session).source_location }
+
+    it "is now defined in the patch file" do
+      expect(source_location.first).to eq(patch_filepath)
+    end
+
+    # If this spec fails, make sure to check the difference between the patched
+    # code and what is in the original gem, and make the necessary changes so
+    # the only changes are to the http_proxy.user and http_proxy.password
+    # lines.
+    it "only changes the necessary lines" do
+      # We can basically assume with our current patch that the line counts for
+      # the method will be the same, so don't try and lex the file to get the
+      # method definition, just calculate it from a known quantity (from patch)
+      #
+      #   * patch_code is determined by taking the source_location line and
+      #     reading up to the 5th to last line in the file.
+      #   * original_last_lineno is determined by determining taking the line
+      #     number from the @_original_start_session_source_loc, and reading
+      #     the number of lines from the original gem source equal to the total
+      #     lines from the `patch_code`
+      #
+      # If the number of lines is off between the patch_code and the original,
+      # then it will fail, and that is also acceptable
+      patch_code = File.readlines(patch_filepath)[source_location[1] - 1..-5].join("")
+
+      original_file, original_lineno = described_class.instance_variable_get(:@_original_start_session_source_loc)
+      original_last_lineno = original_lineno + patch_code.lines.count - 1
+      original_code = File.readlines(original_file)[original_lineno - 1..original_last_lineno - 1].join("")
+
+      # Expect lines 1 through 9 to be the same in both code paths
+      expect(patch_code.lines[0..8].join).to   eq(original_code.lines[0..8].join)
```

Comment:
Does this look easy to understand if this test fails? This is a clever way to check if the patch is needed.  I'm afraid if the test failures aren't easy to decipher, we won't know what's happening when/if we're running against a newer aws-sdk that changed these source lines.

Reply:
_two weeks later..._ 

OH-EM-GEE!!  A response!

* * *

Sorry... Yeah, it isn't too bad.

Here is an example, isn't exactly what you would get, but it was an example of one of the errors I had when working things out:

```console
$ bundle exec rspec spec/lib/patches/aws-sdk-core/seahorse_client_net_http_pool_patch_spec.rb

Randomized with seed 19001
F.

Failures:

  1) Seahorse::Client::NetHttp::ConnectionPool#start_session (monkey patched) only changes the necessary lines
     Failure/Error: expect(patch_code).to eq(original_code)

       expected: "        def start_session endpoint\n\n          endpoint = URI.parse(endpoint)\n\n          args = [...       http.use_ssl = false\n          end\n\n          http.start\n          http\n        end\n\n"
            got: "        def start_session endpoint\n\n          endpoint = URI.parse(endpoint)\n\n          args = [...         http.use_ssl = false\n          end\n\n          http.start\n          http\n        end\n"

       (compared using ==)

       Diff:
       @@ -7,8 +7,8 @@
                  args << endpoint.port
                  args << http_proxy.host
                  args << http_proxy.port
       -          args << http_proxy.user
       -          args << http_proxy.password
       +          args << (http_proxy.user && CGI::unescape(http_proxy.user))
       +          args << (http_proxy.password && CGI::unescape(http_proxy.password))

                  http = ExtendedSession.new(Net::HTTP.new(*args.compact))
                  http.set_debug_output(logger) if http_wire_trace?

     # ./spec/lib/patches/aws-sdk-core/seahorse_client_net_http_pool_patch_spec.rb:31:in `block (3 levels) in <top (required)>'

Finished in 0.02515 seconds (files took 6.97 seconds to load)
2 examples, 1 failure

Failed examples:

rspec ./spec/lib/patches/aws-sdk-core/seahorse_client_net_http_pool_patch_spec.rb:18 # Seahorse::Client::NetHttp::ConnectionPool#start_session (monkey patched) only changes the necessary lines

Randomized with seed 19001
```

Reply:
I actually happened to look over this a bit earlier, and the comments I left were probably not the best as well.  Any suggestions in that department?

Reply:
That is fair.  I will look into trying to introduce the "why" into it first, and then go on to explain the "how" as a follow up (for the nerds the actually read my 📚 ...).

Reply:
haha, yeah, I try to pretend I'm writing to future me at 5pm on a Friday... rarely do you need to explain what the code is doing unless it's truly complicated... usually you just want to know why it's doing this or that

Reply:
@jrafanie Sorry this took a while, but I think I updated the comments to make them a little bit easier to understand, and started off with the "why" as well.  Left the "how" in there as well for the :neckbeard: crowd.